src/Swift.Bindings/src/Parser/SwiftABIParser.cs

```diff
@@ -284,15 +284,11 @@ private List<BaseDecl> CollectDeclarations(IEnumerable<Node> nodes, BaseDecl par
                     metadataPtr = DynamicLibraryLoader.invoke(_dylibPath, GetMetadataAccessor(node));
                     var swiftTypeInfo = new SwiftTypeInfo { MetadataPtr = metadataPtr };
 
-                    if (node.DeclAttributes != null && Array.IndexOf(node.DeclAttributes, "Frozen") != -1 && 
-                        (!swiftTypeInfo.ValueWitnessTable->IsNonPOD || !swiftTypeInfo.ValueWitnessTable->IsNonBitwiseTakable))
-                    {
-                        decl = CreateStructDecl(node, parentDecl, moduleDecl);
-                    }
-                    else
-                    {
-                        decl = CreateClassDecl(node, parentDecl, moduleDecl);
-                    }
+                    // TODO: Find better place for this
+                    typeRecord.IsBlittable = !swiftTypeInfo.ValueWitnessTable->IsNonPOD || !swiftTypeInfo.ValueWitnessTable->IsNonBitwiseTakable; //TODO: This is not the full picture.
```

Comment:
Deciding on whether a struct should be projected as a class or struct on c# side is more complex than that. We will have to implement a correct algorithm for that. I guess for now the question is where this should live. ABIParsing?

Reply:
I agree that metadata info should live in the type database. I suggest to retrieve it using MachO reader, but I think it is important to have it at parsing time. Do we want to read .dylib before the ABI parsing to consume information for projections from type database?

Reply:
I agree that this information should end up in `TypeDatabase`. But I think it also needs to be present on a `TypeDecl`, otherwise  handling each `TypeDecl`  would require reading `TypeDatabase` entry.

>Do we want to read .dylib before the ABI parsing to consume information for projections from type database?
I am not sure I understand

Reply:
Populate type database using MachO reader before ABI parsing, so metadata information can be consumed without calling into dylib from `HandleTypeDecl`.

Reply:
I was looking into how to do this and it's a little more complicated than you think. The problem is that the value witness table for any particular type might not live in the dylib that you're reading. For example, if I define a swift type like this:
```swift
public struct NoInThisDylib {
    public let contents: Int
    public init(with: Int) {
        contents = with
    }
}
```
Then the value witness table for this type is an external reference to the value witness table for Int64 from libswiftCore.dylib. IIRC all heap allocated types (classes) share the same value witness table.

Reply:
I created the following issues to track the outcomes of the discussion:
https://github.com/dotnet/runtimelab/issues/2892
https://github.com/dotnet/runtimelab/issues/2886